In [1]:
import numpy as np

## 1. Cálculo de entropía

In [2]:
# Función para calcular la entropía estimada a partir de las realizaciones
def entropy_estimation(data):
    unique_values, counts = np.unique(data, return_counts=True)
    probabilities = counts / len(data)  # Probabilidades estimadas (frecuencias relativas)
    entropy = -np.sum(probabilities * np.log2(probabilities))  # Cálculo de entropía
    return entropy

### ENTROPIA DE UN TEXTO

In [23]:
texto2 = """En un lugar de la Mancha2, de cuyo nombre no quiero acordarme3, no ha mucho tiempo que vivía un hidalgo de los de lanza en astillero, adarga antigua, rocín flaco y galgo corredor4. Una olla de algo más vaca que carnero, salpicón las más noches5, duelos y quebrantos los sábados6, lantejas los viernes7, algún palomino de añadidura los domingos8, consumían las tres partes de su hacienda9. El resto della concluían sayo de velarte10, calzas de velludo para las fiestas, con sus pantuflos de lo mesmo11, y los días de entresemana se honraba con su vellorí de lo más fino12. Tenía en su casa una ama que pasaba de los cuarenta y una sobrina que no llegaba a los veinte, y un mozo de campo y plaza que así ensillaba el rocín como tomaba la podadera13. Frisaba la edad de nuestro hidalgo con los cincuenta años14. Era de complexión recia, seco de carnes, enjuto de rostro15, gran madrugador y amigo de la caza. Quieren decir que tenía el sobrenombre de «Quijada», o «Quesada», que en esto hay alguna diferencia en los autores que deste caso escriben, aunque por conjeturas verisímilesII se deja entender que se llamaba «Quijana»III, 16. Pero esto importa poco a nuestro cuento: basta que en la narración dél no se salga un punto de la verdad.

Es, pues, de saber que este sobredicho hidalgo, los ratos que estaba ocioso —que eran los más del año—, se daba a leer libros de caballerías, con tanta afición y gusto, que olvidó casi de todo punto el ejercicio de la caza y aun la administración de su hacienda; y llegó a tanto su curiosidad y desatino en esto17, que vendió muchas hanegas de tierra de sembradura para comprar libros de caballerías en queIV leer18, y, así, llevó a su casa todos cuantos pudo haber dellos; y, de todos, ningunos le parecían tan bienV como los que compuso el famoso Feliciano de Silva19, porque la claridad de su prosa y aquellas entricadas razones suyas le parecían de perlas, y más cuando llegaba a leer aquellos requiebros y cartas de desafíos20, donde en muchas partes hallaba escrito: «La razón de la sinrazón que a mi razón se hace, de tal manera mi razón enflaquece, que con razón me quejo de la vuestra fermosura»21. Y también cuando leía: «Los altos cielos que de vuestra divinidad divinamente con las estrellas os fortifican y os hacen merecedora del merecimiento que merece la vuestra grandeza...»22

Con estas razones perdía el pobre caballero el juicio, y desvelábase por entenderlas y desentrañarles el sentido, que no se lo sacara ni las entendiera el mesmo Aristóteles, si resucitara para solo ello. No estaba muy bien con las heridas que don Belianís daba y recebía, porque se imaginaba que, por grandes maestros que le hubiesen curado, no dejaría de tener el rostro y todo el cuerpo lleno de cicatrices y señales23. Pero, con todo, alababa en su autor aquel acabar su libro con la promesa de aquella inacabable aventura, y muchas veces le vino deseo de tomar la pluma y dalle fin al pie de la letra como allí se promete24; y sin duda alguna lo hiciera, y aun saliera con ello25, si otros mayores y continuos pensamientos no se lo estorbaran. Tuvo muchas veces competencia con el cura de su lugar —que era hombre docto, graduado en Cigüenza—26 sobre cuál había sido mejor caballero: Palmerín de Ingalaterra o Amadís de Gaula27; mas maese Nicolás, barbero del mesmo pueblo28, decía que ninguno llegaba al Caballero del Febo, y que si alguno se le podía comparar era don Galaor, hermano de Amadís de Gaula, porque tenía muy acomodada condición para todo, que no era caballero melindroso, ni tan llorón como su hermano, y que en lo de la valentía no le iba en zaga29.

En resolución, él se enfrascó tanto en su letura, que se le pasaban las noches leyendo de claro en claro30, y los días de turbio en turbio; y así, del poco dormir y del mucho leer, se le secó el celebro de manera que vino a perder el juicio31. Llenósele la fantasía de todo aquello que leía en los libros, así de encantamentos como de pendencias, batallas, desafíos, heridas, requiebros, amores, tormentas y disparates imposibles; y asentósele de tal modo en la imaginación que era verdad toda aquella máquina de aquellas soñadas invencionesVI que leía32, que para él no había otra historia más cierta en el mundo33. Decía él que el Cid Ruy Díaz había sido muy buen caballero, pero que no tenía que ver con el Caballero de la Ardiente Espada, que de solo un revés había partido por medio dos fieros y descomunales gigantes34. Mejor estaba con Bernardo del Carpio, porque en Roncesvalles había muerto a Roldán, el encantado35, valiéndose de la industria de Hércules, cuando ahogó a AnteoVII, el hijo de la Tierra, entre los brazos36. Decía mucho bien del gigante Morgante, porque, con ser de aquella generación gigantea, que todos son soberbios y descomedidos, él solo era afable y bien criado37. Pero, sobre todos, estaba bien con Reinaldos de Montalbán, y más cuando le veía salir de su castillo y robar cuantos topaba, y cuando en allende robó aquel ídolo de Mahoma que era todo de oro, según dice su historia38. Diera él, por dar una mano de coces al traidor de Galalón39, al ama que tenía, y aun a su sobrina de añadidura."""


# Opción A: Entropía de los caracteres directamente
datos_caracteres = list(texto)
e_caracteres = entropy_estimation(datos_caracteres)

# Opción B: Entropía de los valores ASCII (numérico)
datos_num = [ord(c) for c in texto]
e_num = entropy_estimation(datos_num)

print(f"Entropía calculada: {e_caracteres:.4f} bits por símbolo")
print(f"Entropía calculada: {e_num:.4f} bits por símbolo")

# 2. Preprocesamiento sin 're' (Limpieza manual)
# Convertimos a minúsculas para unificar símbolos (ej: 'E' y 'e')
texto_min = texto.lower()

# Filtramos: solo conservamos letras (a-z) y espacios, eliminando números y puntuación
# Esto actúa como un filtro para eliminar "ruido" impulsivo
caracteres_filtrados = [c for c in texto_min if c.isalpha() or c.isspace()]

# 3. Cálculo de la entropía
resultado = entropy_estimation(caracteres_filtrados)

print(f"Entropía del texto (normalizado): {resultado:.4f} bits/símbolo")

Entropía calculada: 4.2460 bits por símbolo
Entropía calculada: 4.2460 bits por símbolo
Entropía del texto (normalizado): 4.0015 bits/símbolo


### 1.1 Ejemplo

In [24]:
# Realizaciones de las variables aleatorias X e Y
# (reemplazar estos valores con los datos reales de las realizaciones)
X = [1, 2, 1, 3, 2, 2, 3, 1, 1, 2]  # Ejemplo de realizaciones de X
Y = [2, 3, 2, 3, 1, 2, 3, 1, 3, 2]  # Ejemplo de realizaciones de Y
realizaciones = np.array([
    [1, 2],
    [2, 3],
    [1, 2],
    [3, 3],
    [2, 1],
    [2, 2],
    [3, 3],
    [1, 1],
    [1, 3],
    [2, 2]
])

# Calcular H(X) y H(Y) usando la función de estimación de entropía
H_X = entropy_estimation(X)
H_Y = entropy_estimation(Y)

# Mostrar los resultados
print(f"La entropía estimada H(X) es: {H_X:.4f} bits")
print(f"La entropía estimada H(Y) es: {H_Y:.4f} bits")


La entropía estimada H(X) es: 1.5219 bits
La entropía estimada H(Y) es: 1.5219 bits


## 2. Cálculo de entropía conjunta

In [35]:
def calculo_entropia_conjunta(realizaciones):
    # ... (todo el código anterior se mantiene igual) ...
    X = realizaciones[:,0]
    Y = realizaciones[:,1]
    alfabeto_x = np.unique(realizaciones[:,0])
    alfabeto_y = np.unique(realizaciones[:,1])
    tam_x = len(alfabeto_x)
    tam_y = len(alfabeto_y)
    matriz_frec = np.zeros((tam_x, tam_y), dtype=int)

    for xi, yi in zip(X,Y):
        i = np.where(alfabeto_x == xi)[0][0]
        j = np.where(alfabeto_y == yi)[0][0]  
        matriz_frec[i, j] += 1

    matriz_prob = matriz_frec/len(realizaciones)
    entropiaX = entropy_estimation(X)
    entropiaY = entropy_estimation(Y)
    entropiaConjunta = -np.sum(matriz_prob[matriz_prob > 0] * np.log2(matriz_prob[matriz_prob > 0]))

    return alfabeto_x, alfabeto_y, matriz_prob, entropiaX, entropiaY, entropiaConjunta

# Llamada a la función
calculo_entropia_conjunta(realizaciones)

(array([1, 2, 3]),
 array([1, 2, 3]),
 array([[0.1, 0.2, 0.1],
        [0.1, 0.2, 0.1],
        [0. , 0. , 0.2]]),
 np.float64(1.5219280948873621),
 np.float64(1.5219280948873621),
 np.float64(2.7219280948873625))

### 2.1 Ejemplo

In [37]:
# Ejecución de la función y desempaquetado de resultados
alfabeto_x, alfabeto_y, matriz_prob, entropiaX, entropiaY, entropiaConjunta = calculo_entropia_conjunta(realizaciones)

# Visualización de los resultados obtenidos
print("=== RESULTADOS DEL ANÁLISIS DE ENTROPÍA CONJUNTA ===")
print(f"Alfabeto de X: {alfabeto_x}")
print(f"Alfabeto de Y: {alfabeto_y}")
print("\nMatriz de Probabilidad Conjunta P(X, Y):")
print(matriz_prob)
print(f"\nEntropía H(X): {entropiaX:.4f} bits")
print(f"Entropía H(Y): {entropiaY:.4f} bits")
print(f"Entropía Conjunta H(X, Y): {entropiaConjunta:.4f} bits")
print("====================================================")



=== RESULTADOS DEL ANÁLISIS DE ENTROPÍA CONJUNTA ===
Alfabeto de X: [1 2 3]
Alfabeto de Y: [1 2 3]

Matriz de Probabilidad Conjunta P(X, Y):
[[0.1 0.2 0.1]
 [0.1 0.2 0.1]
 [0.  0.  0.2]]

Entropía H(X): 1.5219 bits
Entropía H(Y): 1.5219 bits
Entropía Conjunta H(X, Y): 2.7219 bits


## 3. Cálculo de entropía condicional

In [ ]:
import numpy as np

def calculo_entropia_condicional(realizaciones):
    """
    Calcula el alfabeto de X e Y, las probabilidades condicionales y las entropías 
    condicionales de X dado Y y Y dado X utilizando la lógica de entropía conjunta.
    """
    # 1. Obtener datos base llamando a la función de entropía conjunta anterior
    alfabeto_x, alfabeto_y, matriz_prob, entropiaX, entropiaY, entropiaConjunta = calculo_entropia_conjunta(realizaciones)

    # 2. Calcular las frecuencias (probabilidades) marginales de X e Y
    # Sumar filas para p(x) y columnas para p(y)
    p_X = np.sum(matriz_prob, axis=1)
    p_Y = np.sum(matriz_prob, axis=0)

    # 3. Inicializar matrices para probabilidades condicionales
    X_Y = np.zeros(matriz_prob.shape)
    Y_X = np.zeros(matriz_prob.shape)

    # 4. Calcular las probabilidades condicionales P(X | Y) y P(Y | X)
    # Se aplica la definición: P(A|B) = P(A,B) / P(B)
    for i in range(matriz_prob.shape[0]):      # filas -> X
        for j in range(matriz_prob.shape[1]):  # columnas -> Y
            if p_Y[j] > 0:
                X_Y[i, j] = matriz_prob[i, j] / p_Y[j]
            if p_X[i] > 0:
                Y_X[i, j] = matriz_prob[i, j] / p_X[i]

---


    # 5. Calcular las entropías condicionales H(X | Y) y H(Y | X)
    # H(X|Y) = -Σ Σ p(x,y) * log2( p(x|y) )
    # Usamos matriz_prob para la ponderación externa y X_Y/Y_X para el logaritmo
    mask = matriz_prob > 0
    entropia_X_dado_Y = -np.sum(matriz_prob[mask] * np.log2(X_Y[mask]))
    entropia_Y_dado_X = -np.sum(matriz_prob[mask] * np.log2(Y_X[mask]))

    return alfabeto_x, alfabeto_y, X_Y, Y_X, entropia_X_dado_Y, entropia_Y_dado_X



### 3.1 Ejemplo 

In [41]:
# Ejecución de la función y desempaquetado de resultados condicionales
alfabeto_x, alfabeto_y, prob_XY, prob_YX, entr_XY, entr_YX = calculo_entropia_condicional(realizaciones)

# Visualización de los resultados obtenidos
print("=== RESULTADOS DEL ANÁLISIS DE ENTROPÍA CONDICIONAL ===")
print(f"H(X|Y) [Incertidumbre de X dado Y]: {entr_XY:.4f} bits")
print(f"H(Y|X) [Incertidumbre de Y dado X]: {entr_YX:.4f} bits")

print("\nMatriz de Probabilidad Condicional P(X|Y):")
print(prob_XY)

print("\nMatriz de Probabilidad Condicional P(Y|X):")
print(prob_YX)
print("======================================================")

=== RESULTADOS DEL ANÁLISIS DE ENTROPÍA CONDICIONAL ===
H(X|Y) [Incertidumbre de X dado Y]: 1.2000 bits
H(Y|X) [Incertidumbre de Y dado X]: 1.2000 bits

Matriz de Probabilidad Condicional P(X|Y):
[[0.5  0.5  0.25]
 [0.5  0.5  0.25]
 [0.   0.   0.5 ]]

Matriz de Probabilidad Condicional P(Y|X):
[[0.25 0.5  0.25]
 [0.25 0.5  0.25]
 [0.   0.   1.  ]]


## 4. Cálculo información mutua

In [ ]:
def calculo_informacion_mutua(realizaciones):
    """
    Calcula la información mutua entre dos variables aleatorias discretas.
    
    Parámetros:
        realizaciones (array): Matriz N x 2 con las realizaciones de dos variables aleatorias discretas.
    
    Retorno:
        informacionMutua (float): Escalar que representa la información mutua entre las dos variables.
    """


    alfabeto_x, alfabeto_y, matriz_prob, entropiaX, entropiaY, entropiaConjunta = calculo_entropia_conjunta(realizaciones)

    _, _, prob_XY, prob_YX, entr_XY, entr_YX = calculo_entropia_conjunta(realizaciones)
    
    informacionMutua = entropiaX + entropiaY - entropiaConjunta
    return informacionMutua

calculo_informacion_mutua(realizaciones)


def calculo_informacion_mutua(realizaciones):
    """
    Calcula la información mutua entre dos variables aleatorias discretas
    utilizando la relación: I(X;Y) = H(X) + H(Y) - H(X,Y)
    """
    # 1. Obtenemos las entropías marginales y la conjunta
    _, _, _, hx, hy, h_conjunta = calculo_entropia_conjunta(realizaciones)
    
    # 2. La Información Mutua es la intersección de la información de X e Y
    # I(X;Y) = H(X) + H(Y) - H(X,Y)
    informacionMutua = hx + hy - h_conjunta
    
    return informacionMutua


### 4.1 Ejemplo

In [71]:
# Ejemplo de uso con una matriz de realizaciones

# Calcular la información mutua
informacionMutua = calculo_informacion_mutua(realizaciones)
informacionMutua

np.float64(0.32192809488736174)

## 5. Capacidad canal 

In [44]:
import numpy as np

def capacidad_canal_44(P):
    """
    Calcula la capacidad de un canal de 4x4 y la distribución de entrada óptima
    maximizando la Información Mutua I(X;Y) = H(Y) - H(Y|X).
    """
    capacidad_canal = 0
    pX_optima = np.zeros(4)
    paso = 0.05
    
    # 1. Calcular la entropía condicional H(Y|X)
    # Como H(Y|X) = sum_x p(x) H(Y|X=x), y H(Y|X=x) depende solo de las filas de P.
    # Calculamos la entropía de cada fila de la matriz de transición
    H_Y_dado_X_fijo = np.zeros(4)
    for i in range(4):
        fila = P[i, :]
        fila_valida = fila[fila > 0]
        H_Y_dado_X_fijo[i] = -np.sum(fila_valida * np.log2(fila_valida))
    
    # 2. Recorrer todas las combinaciones posibles de pX para maximizar I(X;Y)
    for p1 in np.arange(0, 1 + paso, paso):
        for p2 in np.arange(0, 1 - p1 + paso, paso):
            for p3 in np.arange(0, 1 - p1 - p2 + paso, paso):
                p4 = 1 - p1 - p2 - p3
                if p4 < -1e-9 or p4 > 1.0000001: # Tolerancia para errores de punto flotante
                    continue
                
                # Vector de probabilidades de entrada p(X)
                pX = np.array([p1, p2, p3, p4])
                
                # 3. Calcular la distribución de salida p(Y) = p(X) * P
                # Esto es la suma p(y) = sum_x p(x) * P(y|x)
                pY = np.dot(pX, P)
                
                # 4. Calcular la entropía de salida H(Y)
                pY_valido = pY[pY > 0]
                HY = -np.sum(pY_valido * np.log2(pY_valido))
                
                # 5. Calcular la entropía condicional total H(Y|X) = sum p(x)H(Y|X=x)
                HYX = np.sum(pX * H_Y_dado_X_fijo)
                
                # 6. Calcular la información mutua I(X;Y)
                informacion_mutua = HY - HYX
                
                # 7. Actualizar si encontramos un máximo (Capacidad = max I(X;Y))
                if informacion_mutua > capacidad_canal:
                    capacidad_canal = informacion_mutua
                    pX_optima = pX

    return capacidad_canal, pX_optima

In [46]:
# Ejemplo de uso con una matriz de canal 4x4
P = np.array([
    [0, 2, 0, 0],
    [5, 1, 0, 5],
    [14, 7, 0, 0],
    [0, 2, 12, 3]
])
# 1. Normalización de la matriz de canal P proporcionada
# Dividimos cada fila por su suma para que sumen 1 (P(Y|X) válida)
P_norm = P / P.sum(axis=1, keepdims=True)

# 2. Ejecución de mi función de capacidad
capacidad, pX_opt = capacidad_canal_44(P_norm)

# 3. Visualización de resultados
print("=== ANÁLISIS DE CAPACIDAD DE CANAL 4x4 ===")
print("\nMatriz de Canal Normalizada P(Y|X):")
print(P_norm)
print(f"\nCapacidad del Canal (C): {capacidad:.4f} bits/símbolo")
print(f"Distribución de Entrada Óptima p(X): {pX_opt}")
print(f"Suma de p(X) óptima: {np.sum(pX_opt):.2f}")
print("===========================================")

=== ANÁLISIS DE CAPACIDAD DE CANAL 4x4 ===

Matriz de Canal Normalizada P(Y|X):
[[0.         1.         0.         0.        ]
 [0.45454545 0.09090909 0.         0.45454545]
 [0.66666667 0.33333333 0.         0.        ]
 [0.         0.11764706 0.70588235 0.17647059]]

Capacidad del Canal (C): 1.0948 bits/símbolo
Distribución de Entrada Óptima p(X): [0.4  0.25 0.1  0.25]
Suma de p(X) óptima: 1.00


In [51]:
# ejemplo texto
import numpy as np

# 1. Preparación del Texto (Limpieza de "ruido" inicial)
texto_quijote = """En un lugar de la Mancha2, de cuyo nombre no quiero acordarme3, 
no ha mucho tiempo que vivía un hidalgo de los de lanza en astillero, 
adarga antigua, rocín flaco y galgo corredor4..."""


# Normalización: minúsculas y solo caracteres alfabéticos (Filtro de "suavizado")
texto_limpio = "".join([c.lower() for c in texto_quijote if c.isalpha() or c.isspace()])

# --- CÁLCULO 1: Entropía de Palabras (Riqueza de la fuente) ---
palabras = texto_limpio.split()
H_palabras = entropy_estimation(palabras)

# --- CÁLCULO 2: Redundancia del Lenguaje ---
# Calculamos la entropía máxima posible para el alfabeto español (aprox. 27 letras + espacio)
H_max = np.log2(28) 
H_caracteres = entropy_estimation(list(texto_limpio))
redundancia = 1 - (H_caracteres / H_max)

# --- CÁLCULO 3: Dependencia entre Letras (Entropía Conjunta y Condicional) ---
# Creamos pares de letras consecutivas (Letra actual X, Letra siguiente Y)
X = list(texto_limpio[:-1])
Y = list(texto_limpio[1:])
realizaciones_texto = np.array(list(zip(X, Y)))

#  relación entre caracteres adyacentes
alf_x, alf_y, m_prob, hx, hy, h_conjunta = calculo_entropia_conjunta(realizaciones_texto)
_, _, _, _, h_x_dado_y, _ = calculo_entropia_condicional(realizaciones_texto)

# --- VISUALIZACIÓN DE RESULTADOS ---
print("=== ANÁLISIS DE INFORMACIÓN DEL TEXTO ===")
print(f"1. Entropía de Palabras (H_p): {H_palabras:.4f} bits/palabra")
print(f"2. Entropía de Caracteres (H_c): {H_caracteres:.4f} bits/símbolo")
print(f"3. Redundancia estimada: {redundancia*100:.2f}%")
print(f"4. Entropía Conjunta H(X,Y) entre letras: {h_conjunta:.4f} bits")
print(f"5. Entropía Condicional H(X|Y): {h_x_dado_y:.4f} bits")
print("==========================================")

=== ANÁLISIS DE INFORMACIÓN DEL TEXTO ===
1. Entropía de Palabras (H_p): 4.6202 bits/palabra
2. Entropía de Caracteres (H_c): 4.0378 bits/símbolo
3. Redundancia estimada: 16.01%
4. Entropía Conjunta H(X,Y) entre letras: 6.2876 bits
5. Entropía Condicional H(X|Y): 2.2484 bits



---

### 1. Entropía de Palabras ($H_p = 4.6202$ bits/palabra)
Este valor representa la incertidumbre promedio por cada palabra del texto.
* **Significado**: Al ser mayor que la de caracteres, confirma que las palabras como unidades de información poseen una "energía" informativa más alta.
* **Relación con señales**: Equivale a analizar los picos de mayor magnitud en un espectro; las palabras son los eventos que realmente definen el contenido del mensaje.

---

### 2. Entropía de Caracteres ($H_c = 4.0378$ bits/símbolo)
Es la cantidad de información promedio que aporta cada letra individualmente.
* **Interpretación**: Un valor de ~4 bits es estándar para el español normalizado. 
* **Importancia del preprocesamiento**: Este valor es relativamente bajo gracias a que eliminamos el "ruido" (números y puntuación), lo que permite que la distribución de probabilidades sea más limpia y refleje la estructura del idioma.

---

### 3. Redundancia Estimada ($16.01\%$)
Indica que el $16\%$ del texto no aporta información nueva, sino que sirve para dar estructura.
* **Análisis**: Esta redundancia actúa como un "margen de seguridad". Es lo que permite que, si el texto sufre interferencias (como el ruido gaussiano o sal y pimienta que vimos en imágenes), todavía seamos capaces de reconstruir el mensaje.

---

### 4. Entropía Conjunta ($H(X,Y) = 6.2876$ bits)
Es la incertidumbre total de observar pares de letras consecutivas.
* **Observación clave**: Observo que $H(X,Y) < H(X) + H(Y)$ (ya que $4.03 + 4.03 = 8.06$).
* **Conclusión**: Esta diferencia confirma que existe una fuerte dependencia entre letras adyacentes. Las letras no aparecen al azar, sino siguiendo patrones, similar a cómo los píxeles de una imagen original guardan correlación con sus vecinos antes de ser corrompidos por ruido.



---

### 5. Entropía Condicional ($H(X|Y) = 2.2484$ bits)
Representa la incertidumbre que queda sobre una letra si ya conocemos la anterior.
* **Resultado**: Al ser significativamente menor que la entropía individual ($2.24 < 4.03$), se demuestra que el contexto reduce la incertidumbre a casi la mitad.
* **Analogía con filtrado**: Conocer la letra anterior actúa como un "filtro predictor". Si la señal es predecible, el error de estimación (la incertidumbre residual) disminuye drásticamente, permitiendo una transmisión más eficiente por el canal.

# PRACTICA 2B

In [ ]:
import scipy.io

# 1. Cargar los archivos .mat
# Estos archivos contienen las realizaciones de nuestras variables X e Y
data_x = scipy.io.loadmat('VariableX.mat')
data_xy = scipy.io.loadmat('VariableX_Y.mat')


# Extraigo los números reales del diccionario
muestras_x = data_x['X']

# Lo convierto en un vector simple para mi función de entropía
vector_x = muestras_x.flatten() 





# 1. Extraer la matriz conjunta del diccionario
realizaciones_conjuntas = data_xy['XY']

# 2. Separar las columnas para cálculos individuales
X_desde_canal = realizaciones_conjuntas[:, 0]
Y_desde_canal = realizaciones_conjuntas[:, 1]



Entropía de la Entrada H(X): 1.9998 bits
Entropía de la Salida H(Y): 1.9998 bits
Incertidumbre que queda H(X|Y): 1.5845 bits


In [65]:
h_x_referencia = entropy_estimation(vector_x)

# =================================================================
# 3. PROCESAMIENTO CASO B: CANAL DE COMUNICACIÓN (10.000 muestras)
# =================================================================
# Extraemos la matriz conjunta (Col 0: Entrada X, Col 1: Salida Y)
realizaciones_conjuntas = data_xy['XY']

X_canal = realizaciones_conjuntas[:, 0]
Y_canal = realizaciones_conjuntas[:, 1]

# Calculamos entropías individuales del sistema del canal
h_x_canal = entropy_estimation(X_canal)
h_y_canal = entropy_estimation(Y_canal)



print("--- ANÁLISIS COMPARATIVO DE ENTROPÍA ---")
print(f"H(X) Fuente Referencia (N=5000):  {h_x_referencia:.4f} bits")
print(f"H(X) Entrada del Canal (N=10000): {h_x_canal:.4f} bits")


print("\n--- RESULTADOS DEL CANAL X -> Y ---")
print(f"Entropía de la Salida H(Y):       {h_y_canal:.4f} bits")

--- ANÁLISIS COMPARATIVO DE ENTROPÍA ---
H(X) Fuente Referencia (N=5000):  1.7703 bits
H(X) Entrada del Canal (N=10000): 1.9998 bits

--- RESULTADOS DEL CANAL X -> Y ---
Entropía de la Salida H(Y):       1.9998 bits


In [ ]:
## enttropia conjunta
realizaciones_conjuntas = data_xy['XY']

# 2. Ejecución del cálculo de entropía conjunta
# Nota: h_x e h_y devueltos aquí corresponden a las 10.000 muestras
alf_x, alf_y, m_prob, h_x_c, h_y_c, h_conjunta = calculo_entropia_conjunta(realizaciones_conjuntas)

# 3. Visualización de los resultados de la unión de X e Y
print("=== RESULTADOS DE ENTROPÍA CONJUNTA H(X,Y) ===")
print(f"Alfabeto de la Entrada (X): {alf_x}")
print(f"Alfabeto de la Salida  (Y): {alf_y}")
print(f"\nEntropía Conjunta H(X,Y): {h_conjunta:.4f} bits")

# 4. Verificación de la relación de independencia
# Si fueran independientes, H(X,Y) sería igual a H(X) + H(Y)
h_suma_marginal = h_x_c + h_y_c
print(f"Suma de entropías marginales H(X)+H(Y): {h_suma_marginal:.4f} bits")
print(f"Diferencia (Información compartida): {h_suma_marginal - h_conjunta:.4f} bits")
print("==============================================")

print("\nMatriz de Probabilidad Conjunta P(X,Y):")
print(m_prob)


=== RESULTADOS DE ENTROPÍA CONJUNTA H(X,Y) ===
Alfabeto de la Entrada (X): [0 1 2 3]
Alfabeto de la Salida  (Y): [0 1 2 3]

Entropía Conjunta H(X,Y): 3.5843 bits
Suma de entropías marginales H(X)+H(Y): 3.9996 bits
Diferencia (Información compartida): 0.4153 bits

Matriz de Probabilidad Conjunta P(X,Y):
[[0.     0.0839 0.0781 0.0825]
 [0.0865 0.     0.0794 0.0835]
 [0.083  0.0854 0.     0.082 ]
 [0.0863 0.0828 0.0866 0.    ]]


In [69]:
# CONDICIONAL
# =================================================================
# CÁLCULO ESPECÍFICO DE ENTROPÍA CONDICIONAL (Fichero VariableX_Y.mat)
# =================================================================

# 1. Recuperamos la matriz de realizaciones (10.000 muestras)
realizaciones_conjuntas = data_xy['XY']

# 2. Ejecución del cálculo de entropía condicional
# Obtenemos las matrices de probabilidad condicional y las entropías resultantes
alf_x, alf_y, p_X_dado_Y, p_Y_dado_X, h_x_dado_y, h_y_dado_x = calculo_entropia_condicional(realizaciones_conjuntas)

# 3. Visualización de los resultados de incertidumbre residual
print("=== RESULTADOS DE ENTROPÍA CONDICIONAL ===")
print(f"H(X|Y) [Incertidumbre de X tras observar Y]: {h_x_dado_y:.4f} bits")
print(f"H(Y|X) [Incertidumbre de Y tras conocer X]:  {h_y_dado_x:.4f} bits")

# 4. Interpretación técnica
print("\n--- ANÁLISIS DEL CANAL ---")
if h_y_dado_x > 0.1:
    print(f"El canal introduce ruido: H(Y|X) = {h_y_dado_x:.4f} bits.")
else:
    print("El canal es casi ideal (muy poco ruido).")

if h_x_dado_y > 0:
    print(f"Existe pérdida de información (equivocación): H(X|Y) = {h_x_dado_y:.4f} bits.")
print("==========================================")

# Opcional: Mostrar una de las matrices condicionales
# print("\nMatriz de Probabilidad Condicional P(Y|X):")
# print(p_Y_dado_X)

=== RESULTADOS DE ENTROPÍA CONDICIONAL ===
H(X|Y) [Incertidumbre de X tras observar Y]: 1.5845 bits
H(Y|X) [Incertidumbre de Y tras conocer X]:  1.5844 bits

--- ANÁLISIS DEL CANAL ---
El canal introduce ruido: H(Y|X) = 1.5844 bits.
Existe pérdida de información (equivocación): H(X|Y) = 1.5845 bits.


In [72]:
realizaciones_conjuntas = data_xy['XY']
i_mutua = calculo_informacion_mutua(realizaciones_conjuntas)

# --- VISUALIZACIÓN DE RESULTADOS ---
print("=== RESULTADO DE INFORMACIÓN MUTUA I(X;Y) ===")
print(f"Información Mutua: {i_mutua:.4f} bits")

# Cálculo de eficiencia (opcional para el examen)
# ¿Qué porcentaje de la información de la entrada X llega realmente a Y?
_, _, _, hx_c, _, _ = calculo_entropia_conjunta(realizaciones_conjuntas)
eficiencia = (i_mutua / hx_c) * 100

print(f"Eficiencia de la transmisión: {eficiencia:.2f}%")
print("==============================================")

=== RESULTADO DE INFORMACIÓN MUTUA I(X;Y) ===
Información Mutua: 0.4153 bits
Eficiencia de la transmisión: 20.77%


Es fundamental recordar que la capacidad es el valor máximo teórico de la información mutua. Mientras que calculo_informacion_mutua nos da la información que atraviesa el canal con los datos que ya tenemos (los 10.000 datos del fichero), la función capacidad_canal_44 nos dice lo mejor que podría llegar a funcionar ese canal si optimizáramos la fuente de entrada.

In [73]:
# =================================================================
# CÁLCULO DE LA CAPACIDAD DEL CANAL (Optimización de la fuente)
# =================================================================

# 1. Obtenemos la matriz de probabilidad condicional P(Y|X) 
# Esta matriz es la que define "cómo es el canal" físicamente.
# La obtenemos de nuestra función de entropía condicional previa.
_, _, _, P_transicion, _, _ = calculo_entropia_condicional(realizaciones_conjuntas)

# 2. Calculamos la capacidad y la distribución de entrada óptima
capacidad, pX_opt = capacidad_canal_44(P_transicion)

# 3. Visualización de resultados de capacidad
print("=== ANÁLISIS DE CAPACIDAD DEL CANAL ===")
print(f"Capacidad Máxima (C): {capacidad:.4f} bits/símbolo")
print(f"Distribución óptima de entrada p(X): {pX_opt}")

# 4. Comparación con la Información Mutua actual
i_mutua_actual = calculo_informacion_mutua(realizaciones_conjuntas)
print(f"\nInformación Mutua actual I(X;Y): {i_mutua_actual:.4f} bits")
print(f"Diferencia (Margen de mejora): {capacidad - i_mutua_actual:.4f} bits")
print("========================================")

=== ANÁLISIS DE CAPACIDAD DEL CANAL ===
Capacidad Máxima (C): 0.4154 bits/símbolo
Distribución óptima de entrada p(X): [0.25 0.25 0.25 0.25]

Información Mutua actual I(X;Y): 0.4153 bits
Diferencia (Margen de mejora): 0.0001 bits
